### ISSUE TITLE
Writing ROI files

### ISSUE THREAD

**Issue by NicolasH2** [davidcsterratt/RImageJROI#17]
Thanks a lot for this package. The one thing I cannot find anywhere is a way to write .roi-files from R, so that we can use them in ImageJ. In my specific case I have a matrix where every cell represents a pixel and the value of the cell denotes the ROI name.

I wrote a function create an object of your "ijroi" class from the data, but I cannot write that into a file that ImageJ would except. Is that something you would be interested working on? It would be the first time I see anyone do that and complete your package (from IJ to R and from R to IJ).

Here's my function so far:

```r
createROI <- function(mx, roi_id){
  info <- as.data.frame(which(mx==roi_id, arr.ind=T))
  colnames(info) <- c("x","y")
  top <- as.integer(min(info$y)-1)
  left <- as.integer(min(info$x)-1)
  bottom <- as.integer(max(info$y))
  right <- as.integer(max(info$x))
  roi <- list(
    version=as.integer(228),
    type=as.integer(7),
    top=top,
    left=left,
    bottom=bottom,
    right=right,
    width=right-left,
    height=bottom-top,
    n=as.integer(nrow(info)),
    x1=0,y1=0,x2=0,y2=0,
    strokeWidth=as.integer(0),shapeRoiSize=as.integer(0),strokeColor=as.integer(0),fillColor=as.integer(0),subtype=as.integer(0),
    options=as.integer(0),style=as.integer(0),headSize=as.integer(0),arcSize=as.integer(0),position=as.integer(0),
    name=as.character(roi_id),
    coords=as.matrix(info),
    types=list(polygon=0,rect=1,oval=2,line=3,freeline=4,polyline=5,noRoi=6,freehand=7,traced=8,angle=9,point=10),
    strType='traced',
    xrange=c(left,right),
    yrange=c(top,bottom)
  )
  attr(roi, "class") <- "ijroi"
  return(roi)
}
```

**Comment by davidcsterratt** [comment]
If someone wanted to write the code to create ROI files from ROI objects, I would be interested in incorporating it. I think you would need to essentially convert read.ijroi() into write.ijroi() by using  writeBin() to create functions putFloat(), putShort() etc.

**Comment by davidcsterratt** [comment]
Output of devtools::check_win_release() (see below) means we need to fix some things before uploading to CRAN. @gaole2019 Would you have time to look into these problems?

If so, after you think you've fixed them, please run `devtools::check_win_devel()` (you'll need to supply your email address in response to the prompts) and check that all the errors are fixed in the output that's sent to you by email. Then it would be great if you could create a pull request, with commit messages starting "Issue #17: Writing ROI file fix CRAN checks".

I think the URL problem may be that it should be https, not http.

```
* using log directory 'd:/RCompile/CRANguest/R-release/RImageJROI.Rcheck'
* using R version 4.4.1 (2024-06-14 ucrt)
* using platform: x86_64-w64-mingw32
* R was compiled by
    gcc.exe (GCC) 13.2.0
    GNU Fortran (GCC) 13.2.0
* running under: Windows Server 2022 x64 (build 20348)
* using session charset: UTF-8
* checking for file 'RImageJROI/DESCRIPTION' ... OK
* this is package 'RImageJROI' version '0.1.2'
* package encoding: UTF-8
* checking CRAN incoming feasibility ... WARNING
Maintainer: 'David C Sterratt <david.c.sterratt@ed.ac.uk>'

Insufficient package version (submitted: 0.1.2, existing: 0.1.2)

Found the following (possibly) invalid URLs:
  URL: http://imagej.nih.gov/ij/
    From: DESCRIPTION
    Status: Error
    Message: Could not resolve host: imagej.nih.govij

The Title field should be in title case. Current version is:
'Read and write 'ImageJ' Region of Interest (ROI) files'
In title case that is:
'Read and Write 'ImageJ' Region of Interest (ROI) Files'

The Date field is over a month old.
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking serialization versions ... OK
* checking whether package 'RImageJROI' can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking for future file timestamps ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking code files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking loading without being on the library search path ... OK
* checking use of S3 registration ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking examples ... OK
* checking for unstated dependencies in 'tests' ... OK
* checking tests ... ERROR
  Running 'testthat.R'
Running the tests in 'tests/testthat.R' failed.
Complete output:
  > library(testthat)
  > 
  > test_check("RImageJROI")
  Loading required package: RImageJROI
  [ FAIL 1 | WARN 1 | SKIP 0 | PASS 17 ]
  
  ══ Failed tests ════════════════════════════════════════════════════════════════
  ── Error ('test-writeijroi.R:12:5'): write.ijroi() produces files that are consistent with the structure written ──
  Error in `base::file(file, "wb")`: cannot open the connection
  Backtrace:
      ▆
   1. └─RImageJROI::write.ijroi(...) at test-writeijroi.R:12:5
   2.   └─base::file(file, "wb")
  
  [ FAIL 1 | WARN 1 | SKIP 0 | PASS 17 ]
  Error: Test failures
  Execution halted
* checking PDF version of manual ... [13s] OK
* checking HTML version of manual ... OK
* DONE
Status: 1 ERROR, 1 WARNING
```

**Comment by gaole2019** [comment]
@davidcsterratt Hi, David, I have fixed errors and warnings reported in automated tests and created a pull request (Pull #21 ). There remained one NOTE in the automated test and I do not know how to fix it.  `*checking Rd cross-references ... NOTE
Found the following Rd file(s) with Rd \link{} targets missing package
anchors:
  RImageJROI.Rd: spatstat.geom
  ij2spatstat.Rd: owin, ppp, psp
Please provide package anchors for all Rd \link{} targets not in the
package itself and the base packages.`

**Comment by davidcsterratt** [comment]
Many thanks @gaole2019 . I've managed to fix those problems (see http://127.0.0.1:14145/doc/manual/R-exts.html#Cross_002dreferences and https://github.com/davidcsterratt/RImageJROI/commit/48542cc9ee6d5130e2e890b6fc4cf7433db61fa5) and a few other issues to do with CRAN checks. The updated version of ImageJROI that can write ROI files is now on its way to CRAN.
